plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt

```diff
@@ -261,12 +277,94 @@ class SerializableCodegenImpl(
             return 0 to propStartVar
         } else {
             val superProps = bindingContext.serializablePropertiesFor(superClass).serializableProperties
-            val creator = buildInternalConstructorDesc(propStartVar, 1, classCodegen, superProps)
+            val creator = buildInternalConstructorDesc(descVar, propStartVar, maskVar, classCodegen, superProps)
             invokespecial(superType, "<init>", creator, false)
             return superProps.size to propStartVar + superProps.sumBy { it.asmType.size }
         }
     }
 
+    private fun InstructionAdapter.generateNewGoldenMaskCheck(descVar: Int, maskVar: Int) {
+        val allPresentsLabel = Label()
+        val maskSlotCount = properties.serializableProperties.bitMaskSlotCount()
+        if (maskSlotCount == 1) {
+            var goldenMask = 0
+            var requiredBit = 1
+            for (property in properties.serializableProperties) {
+                if (!property.optional) {
+                    goldenMask = goldenMask or requiredBit
+                }
+                requiredBit = requiredBit shl 1
+            }
+
+            iconst(goldenMask)
+            dup()
+            load(maskVar, OPT_MASK_TYPE)
+            and(OPT_MASK_TYPE)
+            ificmpeq(allPresentsLabel)
+
+            load(maskVar, OPT_MASK_TYPE)
+            iconst(goldenMask)
+            load(descVar, descType)
+            invokestatic(
+                "kotlinx/serialization/internal/$PLUGIN_UTILS_LIBRARY_CLASS", THROW_MISSING_FIELD_EXCEPTION_METHOD,
+                "(IILkotlinx/serialization/descriptors/$SERIAL_DESCRIPTOR_CLASS;)V", false
+            )
+        } else {
+            val fieldsMissingLabel = Label()
+
+            val goldenMaskArray = getGoldenMaskArray()
+            for (i in goldenMaskArray.indices) {
+                val goldenMask = goldenMaskArray[i]
+                val maskIndex = maskVar + i
+                iconst(goldenMask)
+                dup()
+                load(maskIndex, OPT_MASK_TYPE)
+                and(OPT_MASK_TYPE)
+                ificmpne(fieldsMissingLabel)
+            }
+            goTo(allPresentsLabel)
+            //prepare seen array
+            visitLabel(fieldsMissingLabel)
+            iconst(maskSlotCount)
```

Comment:
Maybe you'll find `InstructionAdapter.fillArray` function from JVMCodegenUtil.kt useful